source/adapters/hip/enqueue.cpp

```diff
@@ -1618,25 +1618,47 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueUSMMemcpy2D(
     hipPointerAttribute_t srcAttribs{};
     hipPointerAttribute_t dstAttribs{};
 
+    // Determine if pSrc and/or pDst are system allocated pageable host memory.
     bool srcIsSystemAlloc{false};
     bool dstIsSystemAlloc{false};
-
-    hipError_t hipRes{};
-    // hipErrorInvalidValue returned from hipPointerGetAttributes for a non-null
-    // pointer refers to an OS-allocation, hence pageable host memory. However,
-    // this means we cannot rely on the attributes result, hence we mark system
-    // pageable memory allocation manually as host memory. The HIP runtime can
-    // handle the registering/unregistering of the memory as long as the right
-    // copy-kind (direction) is provided to hipMemcpy2DAsync for this case.
-    hipRes = hipPointerGetAttributes(&srcAttribs, (const void *)pSrc);
-    if (hipRes == hipErrorInvalidValue && pSrc)
+    // hipErrorInvalidValue or hipErrorMemoryAllocation returned from
+    // hipPointerGetAttributes, for a non-null pointer, refers to an
+    // OS-allocation, hence pageable host memory. However, this means we cannot
+    // rely on any attribute result, hence we mark it as system allocated
+    // pageable host memory. The HIP runtime can handle the
+    // registering/unregistering of the memory as long as the right copy-kind
+    // (direction) is provided to hipMemcpy2DAsync for this case.
+    if (pSrc &&
+        hipPointerGetAttributes(&srcAttribs, (const void *)pSrc) != hipSuccess)
```

Comment:
Should we be checking that the return error is actually `hipErrorInvalidValue` or `hipErrorMemoryAllocation`, instead of just checking if it's not `hipSuccess`?

Reply:
@hdelan That condition was actually wrong for ROCm 6, so thanks for checking!!

Also, I found that `hipErrorMemoryAllocation` is irrelevant to figuring out unregistered memory (it is for invalid obtained device memory handle). Since ROCm 6 there is now a new value for the hipMemoryType attribute which can relay that information accurately - `hipMemoryTypeUnregistered` which equals `0` and I was originally confused as I assumed `0` was still refering to `hipMemoryTypeHost` but it appears no more.

The new changes introduced to this PR fix pointer type detections for both this copy2d workaround and the sycl allocation info pointer-type query. Would you mind re-reviewing this again, please? Apologies for that as the new changes are sort of a rework. Thank you!